services/typescript-service.ts

```diff
@@ -0,0 +1,316 @@
+			let projectFiles = this.$fs.enumerateFilesInDirectorySync(projectDir,
+				(fileName: string, fstat: IFsStats) => fileName !== rootNodeModules);
+			let typeScriptFiles = _.filter(projectFiles, file => path.extname(file) === ".ts");
+			let definitionFiles = _.filter(typeScriptFiles, file => _.endsWith(file, ".d.ts"));
+			return { definitionFiles: definitionFiles, typeScriptFiles: typeScriptFiles };
+		}).future<ITypeScriptFiles>()();
+	}
+
+	public isTypeScriptProject(projectDir: string): IFuture<boolean> {
+		return ((): boolean => {
+			let typeScriptFiles = this.getTypeScriptFiles(projectDir).wait();
+
+			if (typeScriptFiles.typeScriptFiles.length > typeScriptFiles.definitionFiles.length) { // We need this check because some of non-typescript templates(for example KendoUI.Strip) contain typescript definition files
+				return true;
+			}
+
+			return false;
+		}).future<boolean>()();
+	}
+
+	private hasTsConfigFile(projectDir: string): IFuture<boolean> {
+		return this.$fs.exists(this.getPathToTsConfigFile(projectDir));
+	}
+
+	private getPathToTsConfigFile(projectDir: string): string {
+		return path.join(projectDir, "tsconfig.json");
+	}
+
+	private getCompilerOptions(projectDir: string, options: ITypeScriptTranspileOptions): IFuture<ITypeScriptCompilerOptions> {
+		return ((): ITypeScriptCompilerOptions => {
+			let tsConfigFile: ITypeScriptConfig;
+			let pathToConfigJsonFile = this.getPathToTsConfigFile(projectDir);
+
+			if (this.hasTsConfigFile(projectDir).wait()) {
+				tsConfigFile = this.$fs.readJson(pathToConfigJsonFile).wait();
+			}
+
+			tsConfigFile = tsConfigFile || { compilerOptions: {} };
+			let compilerOptions = options.compilerOptions || {};
+			let defaultOptions = options.defaultCompilerOptions || {};
+
+			let compilerOptionsKeys = _.union(_.keys(compilerOptions), _.keys(tsConfigFile.compilerOptions), _.keys(defaultOptions));
+
+			let result: ITypeScriptCompilerOptions = {};
+			_.each(compilerOptionsKeys, (key: string) => {
+				// The order here is important.
+				result[key] = compilerOptions[key] || tsConfigFile.compilerOptions[key] || defaultOptions[key];
+			});
+
+			result.noEmitOnError = result.noEmitOnError || false;
+
+			return result;
+		}).future<ITypeScriptCompilerOptions>()();
+	}
+
+	private getTypeScriptCompiler(): IFuture<ITypeScriptCompilerSettings> {
+		return ((): ITypeScriptCompilerSettings => {
+			let tempTscDirectory: string;
+			if (!this.hasInstalledTsc) {
+				tempTscDirectory = this.createTempDirectoryForTsc().wait();
+				let pluginToInstall: INpmDependency = {
+					name: TypeScriptService.TYPESCRIPT_MODULE_NAME,
+					version: TypeScriptService.DEFAULT_TSC_VERSION,
+					installTypes: false
+				};
+				this.$npmService.install(tempTscDirectory, pluginToInstall).wait();
+				this.hasInstalledTsc = true;
+			}
+
+			// Get the path to tsc
+			let typeScriptModuleFilePath = path.join(tempTscDirectory, "node_modules", TypeScriptService.TYPESCRIPT_MODULE_NAME);
```

Comment:
In case you do not enter the `if` clause above (`this.hasInstalledTsc` is equal to `true`) `tempTscDirectory` would remain undefined and you'd get an error when trying to `path.join`